Initial yarp-sys (Rust bindings)

turboladen opened the pull request:
Per [discussion 1068](https://github.com/ruby/yarp/discussions/1068), here's my first attempt at a `yarp-sys` crate: Rust bindings to the YARP C API. It's my first time using the [`bindgen`](https://github.com/rust-lang/rust-bindgen) crate (I've only ever hand-rolled my own Rust FFI stuff), so it's possible I've missed a thing or two there. I've added tests to `rust/yarp-sys/tests` that just attempt to validate that the bindings are getting generated ok. Outside of those tests, there's not really much to the crate.

I've started using these in a separate, higher level crate, and so far they seem to be working alright. Let me know if you'd like more docs or tests or anything else.

---

Questions:

- [x] Add a CI job for running tests

turboladen commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,180 @@
+/// This method only generates code in memory here--it doesn't write it to file.
+///
+fn generate_bindings(ruby_include_path: &Path) -> bindgen::Bindings {
+    bindgen::Builder::default()
```
My approach here was to only expose things in the public C interface. If you don't do any of the `allowlist_*`/`blocklist_*` stuff, `bindgen` will do bindings for everything it finds in the headers you give it.

turboladen commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,180 @@
+        .clang_arg(format!("-I{}", ruby_include_path.display()))
+        .parse_callbacks(Box::new(bindgen::CargoCallbacks))
+        // Structs
+        .allowlist_type("yp_buffer_t") // TEST ✅
```
- [ ] Remove `// Test` (those were just notes for me during development)

turboladen commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,180 @@
+        .allowlist_type("yp_string_list_t") // TEST ✅
+        .allowlist_type("yp_token_type_t")
+        // Enums
+        .rustified_non_exhaustive_enum("yp_comment_type_t") // TEST ✅
```
I don't know if I feel strongly about this choice for `enum`s here, but for my part, I find it more ergonomic to deal with Rust-style enums when working in Rust (as opposed to constants that have integer values), hence the choice for `rustified_`.

`bindgen` info on enum options: https://docs.rs/bindgen/latest/bindgen/struct.Builder.html#enums

---

I went with the `non_exhaustive` enum here because a) I don't know what's planned for `yarp` with regards to these types, and b) `yarp` is still relatively young, so maybe these C enums will gain variants over time? By going with non-exhaustive Rust enums, this forces downstream crates to handle variants that might get added in the future--effectively ensuring forward compatibility if more enum variants are added (adding variants to a standard/exhaustive Rust enum is a breaking change since consumers' code won't handle a new variant and thus won't compile). If we know that the C enums are set in stone, then it'd be nice to change this codegen code to generate exhaustive Rust enums (since those are nicer to deal with in Rust).

tl;dr Going this route takes away a tad bit of ergonomics for consumers, but ensures forward compatibility.

More info on non-exhaustive enums: https://docs.rs/bindgen/latest/bindgen/struct.Builder.html#enums

ianks replied:
Since Ruby syntax is a moving target, I think non-exhaustive default makes sense for most things.

turboladen commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,180 @@
+        .rustified_non_exhaustive_enum("yp_token_type")
+        .rustified_non_exhaustive_enum("yp_unescape_type_t") // TEST ✅
+        // Functions
+        .allowlist_function("yp_buffer_init") // TEST ✅
```
Again, the whitelisted/allowlisted functions here should be all the `YP_EXPORTED_FUNCTION`s; that was my intent, at least. I'm happy to trim it down or add to it if that'd be better.

turboladen commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,180 @@
+        .allowlist_function("yp_unescape_manipulate_string") // TEST ✅
+        .allowlist_function("yp_version") // TEST ✅
+        // Vars
+        .allowlist_var("yp_encoding_ascii")
```
These vars get turned into `static` vars, which seemed nice to have if you're doing anything with those encoding callbacks, but again I'm not sure if they should be in or not.
<img width="253" alt="image" src="https://github.com/ruby/yarp/assets/142010/b1639704-7bfe-49eb-a283-526188965d27">

turboladen replied:
[ ] - Look into using a regex for these (to capture newly added encodings in the future

turboladen commented on rust/yarp-sys/tests/parser_tests.rs:
```diff
@@ -0,0 +1,217 @@
+// Because this crate is effectively generated code, our testing really just needs to check that the
+// codegen process didn't do something wonky (which would probably be a bug in the `bindgen` crate).
+// As such, we just need to verify that we can use each of the types and functions we care
+// about--not necessarily anything about their functionality (since that's all part of C-yarp).
+
```
I'd added this before adding other test files; it's more or less in the README now, so...

- [ ] Remove this comment

turboladen commented on rust/yarp-sys/tests/serialize_tests.rs:
```diff
@@ -0,0 +1,100 @@
+        let serialized = std::slice::from_raw_parts(buffer.value as *mut u8, buffer.length);
+
+        assert_eq!(&serialized[0..4], b"YARP");
+        assert_eq!(serialized[4..5][0], 0); // YP_VERSION_MAJOR
+        assert_eq!(serialized[5..6][0], 4); // YP_VERSION_MINOR
+        assert_eq!(serialized[6..7][0], 0); // YP_VERSION_PATCH
```
I'm a bit iffy on these checks since they'll break when the version changes--thoughts?

(also, isn't yarp on 0.5.0, not 0.4.0?)

turboladen replied:
> Might be worth testing it against CARGO_PKG_VERSION or something?

Hmm... I'm not sure I follow. If I understand correctly, the version info there is inserted by `yarp` (C); where would the yarp-sys crate version come into play here?

turboladen commented on rust/yarp-sys/tests/utils_tests.rs:
```diff
@@ -0,0 +1,131 @@
+        CStr::from_ptr(version)
+    };
+
+    assert_eq!(&cstring.to_string_lossy(), "0.4.0");
```
(same question as in another file--shouldn't I be getting 0.5.0? I see `EXPECTED_YARP_VERSION` in `extension.h` is set to "0.4.0", but I thought the gem was at 0.5.0, no? Are these intended to track each other?)

kddnewton commented:
Thanks @turboladen! I've asked a couple of folks from our team to take a look at this PR. It's going to be in limbo for a bit because I'm going on paternity leave for 4 weeks, but it's going to stay open because once we discuss everything I think it'd be good to merge.

turboladen commented:
> Thanks @turboladen! I've asked a couple of folks from our team to take a look at this PR. It's going to be in limbo for a bit because I'm going on paternity leave for 4 weeks, but it's going to stay open because once we discuss everything I think it'd be good to merge.

Sounds good! Enjoy your time off!

ianks commented:
@turboladen I totally missed this PR when I implemented https://github.com/ruby/yarp/pull/1104

Would prefer if yours got merged though since you put in in some great effort here.

Would you be OK with making the output types, etc match the format from #1104?

turboladen commented:
> Would you be OK with making the output types, etc match the format from #1104?

Hm... could you be a bit more specific? Are you talking about _all_ the structs, enums, functions, typedefs, etc? If that's what you mean, I'd think it'd be less effort to just take your PR instead of this one (since you put a lot off effort into that!).

--- 

It looks like #1104 does work to remove things like `yp_` and `_t` and to change the cases of enums and such. ...which I get, but in my experience with doing FFI things, if I'm a consumer of the FFI library (i.e. a consumer of `yarp-sys`), I find it a lot easier to use if the bindings exactly/closely match the C API. If/when you have to look at the C source or docs to figure out some behavior or something, there's mental overhead to translate, say `parser`, to `yp_parser_t`, as opposed to if the FFI type is exactly the same as the C type; which it's maybe no big deal if you're just looking for 1 thing, but if you get down a rabbit hole, there just becomes more you have to keep on that stack in your brain. 

Sure, the `bindgen` generated stuff here isn't Rust-y, but I think the whole point in having the bindings is to take them and put a nice Rust API on top of those, and anyone who's doing that is going to need to refer to the C code to make sure they're doing the right thing. ...and IMHO, having those closer to the C code is a plus, not a minus.

I suppose another upside in the unaltered bindings is that there's almost nothing to maintain? IDK how much of an argument that is though since the main point is to get them right.

Thoughts?

ianks commented:
> It looks like #1104 does work to remove things like `yp_` and `_t` and to change the cases of enums and such. ...which I get, but in my experience with doing FFI things, if I'm a consumer of the FFI library (i.e. a consumer of `yarp-sys`), I find it a lot easier to use if the bindings exactly/closely match the C API. If/when you have to look at the C source or docs to figure out some behavior or something, there's mental overhead to translate, say `parser`, to `yp_parser_t`, as opposed to if the FFI type is exactly the same as the C type; which it's maybe no big deal if you're just looking for 1 thing, but if you get down a rabbit hole, there just becomes more you have to keep on that stack in your brain.
> 
> Sure, the `bindgen` generated stuff here isn't Rust-y, but I think the whole point in having the bindings is to take them and put a nice Rust API on top of those, and anyone who's doing that is going to need to refer to the C code to make sure they're doing the right thing. ...and IMHO, having those closer to the C code is a plus, not a minus.
> 
> I suppose another upside in the unaltered bindings is that there's almost nothing to maintain? IDK how much of an argument that is though since the main point is to get them right.
> 
> Thoughts?

I think your logic makes perfect sense _when there is a higher level wrapper library_ which users consume. However, I wasn't confident that someone would end up maintaining that library in the long-term so it seemed like generating something close would be a good middle-ground. The middle ground being:

- Keep all function names the exact same
- Make the struct / enum names Rusty
- Keep original C docs

All of this is to say, if there are plans for a higher level library, then I agree it'd be better to go for the exact C naming like your approach.

Regardless, I trust your call on this one and happy go with whatever way you decide in the end!

turboladen commented:
> However, I wasn't confident that someone would end up maintaining that library in the long-term so it seemed like generating something close would be a good middle-ground.

Ah, yeah, good point on maintainability. It sounds like (per #1068) @kddnewton still wants to assess things before determining if a higher-level lib would also make sense in this repo, but if not here, then I'd think the main point of the bindings would be to enable others to create their own higher-level lib using the bindings that they'd find here. ...in which case I think I'd still argue my point, but I think yours is also valid and would be curious to hear if anyone else had thoughts on the matter.

FWIW, I started plugging away over at [turboladen/yarp-rs](https://github.com/turboladen/yarp-rs), where I'd initially planned to just do a straight, general purpose wrapper, but I've kinda paused to think a bit more about my use case (an LSP implementation) in that I might want to wrap `yarp-sys` in such a way that mixes in some async stuff. I guess I only mention this to say that it's been helpful to have the bindings match the C stuff in that work.

In any case, thanks for your feedback!

ianks commented:
Alright, my opinion is now: let's go with the raw naming (`yp_`) and get this thing shipped! ❤️ 

I think some of the bindgen config I had would be good here (namely the derives). Want to do a quick pass over the options to see what you think? Other than that it all LGTM:

https://github.com/ianks/yarp/blob/be76362aa4803bb02f420be5e8e3130a72402496/rust/yarp-bindgen/src/main.rs#L126-L146

ianks commented on rust/yarp-sys/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+edition = "2021"
+license-file = "../../LICENSE.md"
+repository = "https://github.com/ruby/yarp"
+description = "Rust bindings to Ruby's YARP parsing library"
```
we should add a `links` directive:

```suggestion
description = "Rust bindings to Ruby's YARP parsing library"
links = "yarp"
```

turboladen replied:
You know, there was a reason why I specifically didn't do that, but I can't remember what that was... Probably some misinformation I came across in a post somewhere. In any case, I have `build.rs` linking against `rubyparser` (`yarp/build/librubyparser.a`, to be specific), since that was the only static lib I came across--should I be pointing to something else?? FWIW, on macOS, I get a `yarp/lib/yarp.bundle`, but no `libyarp.dylib` anywhere...

turboladen replied:
Weird... my understanding of `links` must be off. If I add `links = "rubyparser"`, I get:

```
error: could not find native static library `rubyparser`, perhaps an -L flag is missing?
```

...but adding `links = "yarp", results in no build errors. ...but even weirder, I also get no build errors if I add `links = "yarpasdfasdf"`. 🤔

turboladen commented:
> Alright, my opinion is now: let's go with the raw naming (`yp_`) and get this thing shipped! ❤️
> 
> I think some of the bindgen config I had would be good here (namely the derives). Want to do a quick pass over the options to see what you think? Other than that it all LGTM:
> 
> https://github.com/ianks/yarp/blob/be76362aa4803bb02f420be5e8e3130a72402496/rust/yarp-bindgen/src/main.rs#L126-L146

Most of those look good to me; a couple questions though:

1. Why `.disable_header_comment()`? Having the `bindgen` version in the bindings seems benign, and maybe even helpful to someone down the road (idk). I don't think I care personally, just curious.
2. These are on by default--any objections to leaving them out?
   - `size_t_is_usize(true)`
   - `layout_tests(true)`
   - `generate_comments(true)`
3. `.generate_cstr(true)`: this is cool--I hadn't seen it before! It requires rust 1.59.0, so I set `rust-version = "1.59.0"` in `Cargo.toml`.
4. Why `.derive_copy(false)`? I tend to like having this turned on...
5. `.clang_arg("-fparse-all-comments")`: This generates warnings that point to the C docs; I'll leave those C comments alone, but just thought I'd point that out.

turboladen commented on rust/yarp-sys/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+repository = "https://github.com/ruby/yarp"
+description = "Rust bindings to Ruby's YARP parsing library"
+links = "yarp"
+authors = ["Steve Loveless <steve.loveless@gmail.com>"]
```
@ianks Can I list you in here?

ianks replied:
Sure!

```suggestion
authors = ["Steve Loveless <steve.loveless@gmail.com>", "Ian Ker-Seymer <hello@ianks.com>"]
```

ianks commented:
> 1. Why `.disable_header_comment()`? Having the `bindgen` version in the bindings seems benign, and maybe even helpful to someone down the road (idk). I don't think I care personally, just curious.

I have the same opinion. Don't care either way.

> 2. These are on by default--any objections to leaving them out?
>    
>    * `size_t_is_usize(true)`
>    * `layout_tests(true)`
>    * `generate_comments(true)`

I tend to prefer keeping them in, since it makes understanding-at-a-glance a bit easier. Up to you though.

> 3. `.generate_cstr(true)`: this is cool--I hadn't seen it before! It requires rust 1.59.0, so I set `rust-version = "1.59.0"` in `Cargo.toml`.

Yeah, we dont need this...

> 4. Why `.derive_copy(false)`? I tend to like having this turned on...

I think for certain types we should derive Copy, but I was more concerned about "freeable" types, such as `Parser` and `Buffer`, etc. Making copy easy on these types could make double frees too easy.

> 5. `.clang_arg("-fparse-all-comments")`: This generates warnings that point to the C docs; I'll leave those C comments alone, but just thought I'd point that out.
> 
> ![image](https://user-images.githubusercontent.com/142010/251571489-3e5fae9a-0b71-47e3-aa14-79ddc300e3a9.png)

Yeah... ideally we wouldn't need this flag, but the comments use `//` comments instead of doxygen style (`/**`)... @kddnewton would be could to convert these at some point.

turboladen commented:
On the bindings builder comment...

1. I left out `.disable_header_comment()`.
2. I added the 3 methods in that are on by default (your point was well taken).
3. I wasn't sure exactly what you mean about the `generate_cstr()` comment--were you saying to remove the `generate_cstr()` call or remove the `ruby-version` setting in `Cargo.toml`?
4. I added the `derive_copy(false)` call', makes sense.
5. Good to know about the docs generating.

ianks commented:
I think this is in a good enough state to ship and start messing around with, thanks so much @turboladen!

@kddnewton Whenever you are ready, free free to hit that :shipit: button!

eregon commented:
This most likely needs a CI check to avoid breaking unknowingly.

turboladen commented:
> This most likely needs a CI check to avoid breaking unknowingly.

Sorry I missed this comment! I just added a new `rust-bindings` workflow.

kddnewton commented:
@turboladen looks like it might just be needing to ignore the files in the manifest check.

turboladen commented:
> @turboladen looks like it might just be needing to ignore the files in the manifest check.

@kddnewton I'm having a bit of a time trying to track this down. I see that the log has:

```
(see /home/runner/work/yarp/yarp/tasks/check_manifest.rake)
```

...but I a) don't see this file in the repo, and b) am not able to grep for anything that looks relevant. Could you point me in the right direction (a bit more :D)? Sorry, I'm not super familiar with Ruby<->C build things. 👀

kddnewton commented:
Hey @turboladen, no problem, it's not exactly intuitive. 

For this, you should add `rust` to the `ignore_directories` array here: https://github.com/ruby/yarp/blob/main/rakelib/check_manifest.rake#L5.

kddnewton commented:
Also if you wouldn't mind squashing the commits, then I think this will basically be good to go!

turboladen commented:
> For this, you should add `rust` to the `ignore_directories` array here: https://github.com/ruby/yarp/blob/main/rakelib/check_manifest.rake#L5.

Interesting... I don't have `rakelib/` in my branch. I suppose that means I need to pull in some changes. :) Thanks--I'll get that in.

turboladen commented:
> Also if you wouldn't mind squashing the commits, then I think this will basically be good to go!

You betcha--once the CI stuff is sorted, I'll squash away and post back.

turboladen commented:
> Also if you wouldn't mind squashing the commits, then I think this will basically be good to go!

Ok, all updated and squashed. If any builds fail, I'll do my best to get those fixed this weekend.

lopopolo commented on .github/workflows/rust-bindings.yml:
```diff
@@ -0,0 +1,39 @@
+          restore-keys: |
+            ${{ runner.os }}-cargo-${{ hashFiles('Cargo.toml') }}
+            ${{ runner.os }}-cargo
+      - name: cargo test
```
GHA workflows for rust crates typically use a matrix strategy and test against multiple rust versions. I see 1.59.0 mentioned as the MSRV in the thread. Testing against that and stable seems useful.

As it is, this workflow tests against whatever version of the rust toolchain happens to be baked into the GHA runner image.

https://github.com/dtolnay/rust-toolchain is a popular action used to setup a Rust toolchain. I use the collection of workflows in https://github.com/artichoke/setup-rust for Artichoke.

turboladen replied:
> GHA workflows for rust crates typically use a matrix strategy and test against multiple rust versions.

I understand that this is common, but is it _really_ typical? I guess IMO, I'd rather just test against current stable (or rather what's on the GH agent) until there's a real need to add the build complexity. And if that means only claiming "we only support Rust version [the latest stable]," then I'm all for that since that gives development flexibility until we decide we want to lock it down. On the other hand, I've only ever done private/company-internal releases, so I don't have OSS+Rust experience and may be off here. I just like the idea of pushing out complexity until it's needed.

> https://github.com/dtolnay/rust-toolchain is a popular action used to setup a Rust toolchain. 

I've never used that; only https://github.com/actions-rs, but whatever's clever. Again, I'd rather keep it simple until there's a need. If anyone else feels strongly about matrix testing, please chime in though and I'll update.

lopopolo replied:
The actions-rs organization and all of its repositories are widely understood to be abandoned: https://github.com/actions-rs/toolchain/issues/216

turboladen replied:
> The actions-rs organization and all of its repositories are widely understood to be abandoned: [actions-rs/toolchain#216](https://github.com/actions-rs/toolchain/issues/216)

I guess not widely enough. :D

lopopolo replied:
😅 sorry about that. the @actions-rs actions have been emitting warnings to the workflow page for upwards of a year due to not being updated for several GitHub Actions security advisories and the pending node 12 runtime removal.

turboladen replied:
haha no worries! Somehow I've managed not to see those warnings. We've used them at work for quite some time for most of our non-trivial Rust projects; I'll have to check those tomorrow now!

lopopolo commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,168 @@
+/// Convenience step for running `rake compile` before doing our stuff.
+///
+fn rake_compile() -> Result<(), io::Error> {
+    let output = Command::new("rake").arg("compile").output()?;
+
+    io::stdout().write_all(&output.stdout)?;
+    io::stderr().write_all(&output.stderr)?;
```
this code can be simplified by invoking the command with `status` instead of `output`. The docs specify that the command will inherit stdout and stderr by default if invoked with either `status` or `spawn`.

- https://doc.rust-lang.org/std/process/struct.Command.html#method.new
- https://doc.rust-lang.org/std/process/struct.Command.html#method.status

turboladen replied:
Nice.

lopopolo commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,168 @@
+}
+
+fn cargo_manifest_path() -> &'static Path {
+    Path::new(env!("CARGO_MANIFEST_DIR"))
```
This env variable is set when the build script is run, not when it is compiled. It must be resolved with `std::env::var_os` at runtime.

https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts

turboladen replied:
Ah, right! Nice catch!

lopopolo commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,168 @@
+    // Add `[root]/build/` to the search paths, so it can find `librubyparser.a`.
+    println!(
+        "cargo:rustc-link-search=native={}",
+        ruby_build_path.display()
```
if the path is not able to be represented as UTF-8, this is unlikely to behave correctly, so prefer to abort.

```suggestion
        ruby_build_path.to_str().unwrap()
```

turboladen replied:
Interesting. I thought the `println!` in build scripts effectively behaves like sending the string via stdin, and in that case you'd want to whatever encoding the OsStr was in. Why is there a need to ensure a UTF-8 string is used here?

lopopolo replied:
`println!` by virtue of taking a `&str` format string requires the fully constructed payload to be UTF-8.

`Path::display` results in a UTF-8 encoded representation of the path, which for paths with invalid UTF-8 contents like `\xFF` or unpaired surrogates on Windows, will not be a lossless transformation (it will include some escaping).

Such paths are unable to be losslessly printed in this way, which means paths containing them won't properly set up the search path.

Rather than delaying this scenario to a more opaque error later in linking, prefer to panic early.

turboladen replied:
Ah ok, that makes sense. I think I read the `Path::display` docs to say that it'd allow non-unicode strings, but didn't catch the lossy part (which certainly makes it sound like it behaves similar to the various `to_string_lossy()` calls in `std`). I'll change. Thanks for the explanation.

lopopolo commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,168 @@
+        .generate_block(true)
+        .generate_comments(true)
+        .generate_cstr(true)
+        .header(ruby_include_path.join("yarp/defines.h").to_string_lossy())
+        .header(ruby_include_path.join("yarp.h").to_string_lossy())
```
similar to my comment above, if the path is not valid UTF-8, the build will fail anyway, so unwrap here.

```suggestion
        .header(ruby_include_path.join("yarp/defines.h").to_str().unwrap())
        .header(ruby_include_path.join("yarp.h").to_str().unwrap())
```

lopopolo replied:
Similar to my comment around `Path::display` above, `to_string_lossy` replaces invalid UTF-8 in paths with the unicode replacement character `\uFFFD`. This is unlikely to result in a successful compilation, so assert that a lossless conversion occurs.

lopopolo commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,168 @@
+        .generate_cstr(true)
+        .header(ruby_include_path.join("yarp/defines.h").to_string_lossy())
+        .header(ruby_include_path.join("yarp.h").to_string_lossy())
+        .clang_arg(format!("-I{}", ruby_include_path.display()))
```
ditto

lopopolo commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,168 @@
+/// Write the bindings to the $OUT_DIR/bindings.rs file. We'll pull these into the actual library
+/// in `src/lib.rs`.
+fn write_bindngs(bindings: bindgen::Bindings) {
+    let out_path = PathBuf::from(std::env::var("OUT_DIR").unwrap());
```
```suggestion
    let out_path = PathBuf::from(std::env::var_os("OUT_DIR").unwrap());
```

turboladen replied:
Similar to my question above, if you want to ensure UTF-8, it'd seem you'd want to use `var()` here, not `var_os()`; why allow non-UTF8 encoding from an env var here, then fail later when you're outputting the `PathBuf` to `stdin`? I'd think if you really want to ensure UTF-8 all around (which I'm still unsure of why you would in these cases), you'd use `var()` and `Path::to_str()`. String encodings are not my expertise, though, so please explain if you have a min!

lopopolo replied:
this one actually is to _avoid_ needing to care about UTF-8 validation. We go from an env var to a pathbuf directly to a `std::fs` API. Using `var_os` here means the entire call chain never once does a UTF-8 validation.

In build scripts and when dealing with path and platform strings in general, my preference is to maintain the native representation for as long as possible, only converting to UTF-8 if there are no other options. Printing to stdout with `println!` and passing to bindgen APIs that take `Into<String>` are such cases where there is no choice but to convert to UTF-8, so we might as well assert that we, in fact, do.

turboladen replied:
Ah yeah, that makes sense. Nice!

lopopolo commented on rust/yarp-sys/src/lib.rs:
```diff
@@ -0,0 +1,36 @@
+//  │ types end up with weird names. I'm sure there's a way to fix that, but I     │
+//  │ haven't been able to. So I'm hand-rolling these for now.                     │
+//  ╰──────────────────────────────────────────────────────────────────────────────╯
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct yp_string_t {
+    pub type_: yp_string_t_type,
+    pub source: *const c_char,
+    pub length: usize,
+}
+
+#[repr(u32)]
+#[non_exhaustive]
+#[derive(Clone, Copy)]
+pub enum yp_string_t_type {
+    YP_STRING_SHARED,
+    YP_STRING_OWNED,
+    YP_STRING_CONSTANT,
+    YP_STRING_MAPPED,
+}
```
can we also derive `Debug` for these types?

turboladen replied:
Yup.

lopopolo commented on rust/yarp-sys/src/lib.rs:
```diff
@@ -0,0 +1,36 @@
+//  │ haven't been able to. So I'm hand-rolling these for now.                     │
+//  ╰──────────────────────────────────────────────────────────────────────────────╯
+#[repr(C)]
+#[derive(Clone, Copy)]
```
can this type derive `Default` to match the bindgen settings?

turboladen replied:
It can, yeah. My only hesitancy is what the default variant should be for `yp_string_t_type`; right now it's `YP_STRING_SHARED`, but does that make sense? I'd almost rather hold off and add this in later after the lib has gotten some use out of it... unless someone familiar with the C API can weigh in as to what's the most common variant of string used?

lopopolo commented on rust/yarp-sys/src/lib.rs:
```diff
@@ -0,0 +1,36 @@
+
+// In `build.rs`, we use `bindgen` to generate bindings based on C headers and `librubyparser`. Here
+// is where we pull in those bindings and make them part of our library.
+include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
```
typically when using `bindgen`, I stick the `include!` in a module so we can scope the `allow` pragmas to only the generated code. Something like this:

```rust
#[allow(non_upper_case_globals)]
#[allow(non_camel_case_types)]
#[allow(non_snake_case)]
mod bindings {
    // In `build.rs`, we use `bindgen` to generate bindings based on C headers and `librubyparser`. Here
    // is where we pull in those bindings and make them part of our library.
    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
}

pub use self::bindings::*;
```

turboladen replied:
I don't really feel that strongly about this either way, although I'd kinda lean towards leaving it as-is since ideally, IMO, this whole crate would be bindgen code. I only have the two manually defined types in there (`yp_string_t` and `yp_string_t_type`) because I don't know how to get bindgen to generate an embedded enum as something other than the via its default naming (`yp_string_t__bindgen_ty_1` in this case) and that doesn't seem ideal to me.

It's been over 6 weeks since I've worked on any of this code, so I'm forgetting what I came across, but IIRC, I came across something that led me to believe that it was (or would be?) doable. If we end up with needing to define type manually for the long term, then yeah, that makes sense to me; otherwise we end up with a `bindings` submodule with no peer modules and that seems unnecessary.

lopopolo replied:
this forces allowing rustc "warn by default" lints for the entire crate, including any unit tests, should they be added in the future.

turboladen replied:
I don't think I see the problem with that, but at this point I'm in favor of getting this merged since I don't have much time to work on this right now (I did 6 weeks ago, but life has changed a bit since then), so I'll just change it.

lopopolo commented on rust/yarp-sys/tests/node_tests.rs:
```diff
@@ -0,0 +1,35 @@
+    let code = CString::new("class Foo; end").unwrap();
+    let mut memsize = MaybeUninit::<yp_memsize_t>::uninit();
+
+    unsafe {
+        yp_parser_init(
+            parser.as_mut_ptr(),
+            code.as_ptr(),
+            code.as_bytes().len(),
+            std::ptr::null(),
+        );
+
+        let parser = parser.assume_init_mut();
+        let node = yp_parse(parser);
+
+        assert_eq!((*node).type_, yp_node_type::YP_NODE_PROGRAM_NODE as u16);
+
+        yp_node_memsize(node, memsize.as_mut_ptr());
+        let memsize = memsize.assume_init();
+        assert_eq!(memsize.memsize, 296);
+        assert_eq!(memsize.node_count, 4);
+
+        yp_node_destroy(parser, node);
+        yp_parser_free(parser);
+    }
```
this is quite lovely! looking forward to using this. 💎

lopopolo commented on rust/yarp-sys/tests/pack_tests.rs:
```diff
@@ -0,0 +1,57 @@
+    let mut encoding_out = MaybeUninit::<yp_pack_encoding>::uninit();
+
+    unsafe {
+        let result = yp_pack_parse(
```
Amazing. Happy to make this obsolete:

- https://github.com/artichoke/artichoke/pull/2605

lopopolo commented on rust/yarp-sys/tests/serialize_tests.rs:
```diff
@@ -0,0 +1,102 @@
+        yp_parser_free(parser);
+
+        // Can't use String -> CString here because `value` contains nul bytes.
+        Vec::from_raw_parts(
+            serialize_buffer.value as *mut u8,
+            serialize_buffer.length,
+            serialize_buffer.capacity,
+        )
```
`Vec::from_raw_parts` can only be safely used on memory that was allocated as a Rust `Vec`. I think you want `slice::from_raw_parts` here.

turboladen replied:
Ah yeah, you're right. At first, I was just making tests for my sake and wasn't planning on committing them, but they do serve as decent examples as to how to use the bindings, so I left them. As you've noted (thank you!), there's some lazy coding in there. Will fix!

lopopolo commented on rust/yarp-sys/tests/serialize_tests.rs:
```diff
@@ -0,0 +1,102 @@
+            metadata,
+        );
+
+        let string = String::from_raw_parts(
```
similarly, this is unsafe if the buffer did not come from a string.

lopopolo commented on rust/yarp-sys/tests/serialize_tests.rs:
```diff
@@ -0,0 +1,102 @@
+            metadata,
+        );
+
+        let string = String::from_raw_parts(
```
this takes ownership of the memory in the `parse_serialize_buffer` and will free it when the `String` is dropped. Is that intentional?

turboladen replied:
Nope.

lopopolo commented on rust/yarp-sys/tests/serialize_tests.rs:
```diff
@@ -0,0 +1,102 @@
+        yp_parser_free(parser);
+
+        // Can't use String -> CString here because `value` contains nul bytes.
+        Vec::from_raw_parts(
```
this takes ownership of the memory in the `serialize_buffer` and will free it when the `Vec` is dropped. Is that intentional?

turboladen replied:
Nope. Switched to use `slice` here.

lopopolo commented on rust/yarp-sys/tests/serialize_tests.rs:
```diff
@@ -0,0 +1,102 @@
+        let buffer = buffer.assume_init_mut();
+        yp_serialize(parser, node, buffer);
+
+        let serialized = std::slice::from_raw_parts(buffer.value as *mut u8, buffer.length);
```
with all of the unsafe code in the tests, I suggest running clippy on this crate. It likely will suggest to replace this `as` cast with `buffer.value.cast::<u8>()` among other things.

lopopolo replied:
this lint in particular that I've called out I believe requires warning on `clippy:pedantic`.

https://rust-lang.github.io/rust-clippy/master/index.html#/ptr_as_ptr

turboladen replied:
Hm... I have my IDE setup to run clippy on every save and it doesn't tell me that. ...and if I run it from the CLI, I get no warnings.

turboladen replied:
TIL... I added `#![warn(clippy::pedantic)]` (and a bunch of others) to `src/lib.rs` and ran `cargo clippy --tests`, and that came back with 0 warnings (after fixing other non-test things). I then ran `cargo clippy --tests -- -W "clippy:pedantic"` and got the warning you mentioned. All these years, I thought adding the `warn` attributes to my crate would apply to testing too.

lopopolo replied:
integration tests (the tests found in the `tests/` directory) are compiled as separate crates and thus must have their own crate-level attributes.

lopopolo commented on rust/yarp-sys/tests/serialize_tests.rs:
```diff
@@ -0,0 +1,102 @@
+        let metadata = std::ptr::null();
+
+        yp_parse_serialize(
+            serialized.as_ptr() as *const i8,
```
this test will fail to compile on ARM linux which defines `c_char` to be `u8`. prefer `serialized.as_ptr().cast::<c_char>()`.

turboladen replied:
TIL. Fixed.

lopopolo replied:
🙇 yea this is a non-obvious one and not one that clippy can help with due to the fact `c_char` is defined as a type alias and thus not present in any form that a lint pass can inspect.

I learned this the hard way too:

- https://github.com/artichoke/artichoke/pull/1632

lopopolo reviewed (commented):
Hi folks, I got an @ on Twitter from @kddnewton. I'm the author of Artichoke Ruby and a likely user for these bindings. I hope my feedback is welcome.

One other thing I'd be interested in seeing set up: the presence of Rust bindings means it would be trivial to fuzz the crate and C parser with `cargo-fuzz`. You can take a peek at these bits for how I've been fuzzing mruby via Artichoke (and have captured some trophies in the process):

- https://github.com/artichoke/artichoke/blob/trunk/.github/workflows/fuzz.yaml
- https://github.com/artichoke/artichoke/tree/trunk/fuzz

lopopolo reviewed (commented):
I'd like to see the CI on this crate expanded to:

- Deny warnings.
- Run clippy.
- Run the tests under leak sanitizer using `rustc`'s `-Z sanitzer=leak` option.
- Run the tests under address sanitizer using `rustc`'s `-Z sanitizer=address` option.

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,168 @@
+        // Vars
+        .allowlist_var(r#"^yp_encoding\S+"#)
+        .generate()
+        .expect("Unable to generate yarp bindings")
```
Seeing this pretty long list of things, including things which are not really public API (like enum), I wonder if it wouldn't be much simpler and safer to use serialization for Rust instead: https://github.com/ruby/yarp/discussions/1068#discussioncomment-6708266

@kddnewton My impression is the C API is not stable yet and it's not public API either, except for the functions marked with `YP_EXPORTED_FUNCTION`. As such it feels wrong that Rust would depend on so many implementation details that are subject to change.

eregon replied:
You can see how little API surface is needed when using serialization at https://github.com/ruby/yarp/pull/1112

lopopolo replied:
I would very much prefer to use the C API, the same one MRI uses, instead of serialization.

turboladen replied:
I would not use this library if it relied on serialization.

eregon replied:
Why not? Could you explain the reasons?

Basically because you depend on internals here you must expect on every release something could break and you'd need to adjust your Rust code using it, even after 1.0, because it's not depending just on the public API (currently just `YP_EXPORTED_FUNCTION`) but on internals too.
And it would be an overhead for YARP maintenance.

For instance, why does this expose `yp_lex_mode_t` to Rust? That seems strictly internal.

I suppose there should be a clear definition of what part of YARP's C API is considered public/fairly stable and what part is internal (and you are on your own if you depend on that).
`YP_EXPORTED_FUNCTION` functions is a start but is too restrictive.
We'd need at least node structs (if there is a field rename there it would break Rust code using it, but seems unavoidable anyway).
Things like parser errors/warnings/comments we'd probably need some functions exposing it cleanly rather than directly the `yp_list_t`.
Basically the fewer structs/enum exposed the better maintenance-wise.

kddnewton replied:
I'm fine with no relying on the serialization API. I do think we can investigate whether templating the structs would be more beneficial to the ergonomics than using bindgen, but we'll need bindgen either way for the functions.

turboladen replied:
> Why not? Could you explain the reasons?

At this point, my only interest is for building a Ruby LSP in Rust. I've used `tree-sitter` and [lib-ruby-parser](https://github.com/lib-ruby-parser/lib-ruby-parser) for some prototypes; both have their pros and cons, but neither are ideal for a Ruby LSP, IMHO. Because requirements of server-side LSP can require some heavy lifting, I really care about eeking out every bit of performance I can. While `yarp` looks amazing for lots of things, I'm still concerned it might not perform in this setting as well as `tree-sitter`, given `tree-sitter`'s explicit features for working with text. I only imagine relying on serialization/deserialization for all of this text would only mean a performance hit, and I'd rather take the complexity and churn if that means I have a Ruby LSP implementation that's not slow.

> Basically because you depend on internals here you must expect on every release something could break and you'd need to adjust your Rust code using it, even after 1.0, because it's not depending just on the public API (currently just YP_EXPORTED_FUNCTION) but on internals too.
And it would be an overhead for YARP maintenance.

I initially pinged @kddnewton in #1068 to see if there was interest; they said "yes", so I made the PR. If there wasn't interest in the code, I'd totally understand that (specifically due to the maintenance), and would happily make these same exact bindings available outside the `ruby` organization (i.e. I'd release them on my own). Like @lopopolo has alluded to, I think that bindings at this level have the potential to be really helpful for others, and if they're part of the `ruby` org, I imagine they'll only receive more attention (which means higher quality) than if they were provided by a lone developer such as myself.

If I've added bindings for anything that's not public, that was an accident, and those things should be removed--specifically due to the argument you're giving. These bindings should only be for the public API, and should cover the least surface area possible while allowing to use that API. If what I've put up is not that, then we should change it.

> Basically the fewer structs/enums exposed the better maintenance-wise.

💯 

> I'm no Rust expert but I would expect it's so much nicer to use if using serialization and template-generated Rust classes (one per node type), because then you don't need to wrap everything in unsafe { }, including node field accesses, Rust owns the memory and knows how to deal with it efficiently, etc.

As I think you could probably tell, at this point there's no code in here that adds any functionality on top of the C API--it effectively just adds the ability to call that API. As I think you're alluding to, to be really useful though, one would want to make another Rust library that builds on these bindings. When building one of these wrapper libs in Rust, the expectation is that you'll always be using `unsafe` and that's fine/expected/the-only-way-to-do-this (I mean, Rust itself uses `unsafe` in its own codebase...). (Sorry if I'm pointing out obvious things--just trying to be thorough) IMO, since those next-level-up wrapper libraries can be opinionated, it could make sense to release just these bindings, then wait for some time (months?) to see what surfaces in the community before introducing "Ruby's official Rust yarp lib" (<- the higher level crate that would abstract away all the `unsafe` and make it nice and ergonomic/Rusty). That'd give time for `yarp` itself to stabilize _and_ give time for these bindings to stabilize, all while providing others (myself included) to use them for their needs. (...and now that I think about it, personally, I don't think I'd need a full-blown wrapper of these bindings--I'd just want to be able to call `parse()`, navigate through the parse tree, and match against `node` variants in that parse tree...)

As far as "Rust owns the memory"... while this is certainly contentious, I really actually like the idea of borrowing/sharing any memory I can from yarp because hopefully that means less copying/cloning/whatever, and hopefully better performance.

eregon replied:
Thank you for the detailed reply.

> I'd rather take the complexity and churn if that means I have a Ruby LSP implementation that's not slow.

I think it is a rather small overhead and I would bet this overhead is not what makes a Ruby LSP slow or fast.
In fact in it was found in  https://github.com/ruby/yarp/pull/1149#issuecomment-1646219164 that the difference is small and also in https://github.com/ruby/yarp/discussions/836#discussioncomment-5960781 (although the Ruby API via C extension copies nodes, which I think is so much simpler and necessary if one must be able to create nodes in Ruby; Also consider this is deserializing in pure Ruby code, it would be so much faster in Rust).
IIRC when I measured parsing+serializing and deserializing in Java for TruffleRuby I found that parsing is about 10x slower than deserializing, IOW deserializing is like a 10% (+ a bit more for serializing, but that should be faster than deserializing as it doesn't need to allocate anything except the buffer) overhead.

> If I've added bindings for anything that's not public, that was an accident, and those things should be removed--specifically due to the argument you're giving. These bindings should only be for the public API, and should cover the least surface area possible while allowing to use that API. If what I've put up is not that, then we should change it.

Alright, then I think probably only parsing (`yp_parse` and its dependencies, including all the node structs) should be exposed to Rust, maybe lexing too if that's useful. @kddnewton mentioned we should probably have a public header and internal header to clarify what is what. That is for sure a useful thing to do and define.

> As far as "Rust owns the memory"... while this is certainly contentious, I really actually like the idea of borrowing/sharing any memory I can from yarp because hopefully that means less copying/cloning/whatever, and hopefully better performance.

I can see why for "read-only" workloads it feels tempting to just borrow the memory.
But if it's to gain 10% is it really worth all the noise in caller code using `unsafe {}` etc?
Also not being able to use node flags easily and all other nice things defined on node classes in Ruby and Java, but not in C because well C doesn't have classes.
I'm genuinely interested how much the performance difference would be and how much easier to use it would be.
I think it'd be worth to try both.

BTW do you really have some code/prototype using this? That could be useful to assess "what needs to be exposed" and most of the rest probably just doesn't need to be exposed.

lopopolo replied:
@eregon you are over indexing on `unsafe` blocks. This is what it means to call extern C APIs in Rust. You should imagine the unsafe blocks around every API call to yarp in MRI. The two usages are equivalent.

As @turboladen explained, it is very unlikely that normal Rust code would directly consume `yarp-sys` and thus would not need to use unsafe blocks. Idiomatic wrappers around foreign code include both the raw bindings (the sys crate) and a higher level safe wrapper which is idiomatic, ergonomic rust with a safe public API.

This crate's sole purpose is to make the public C API available to Rust code. You can think of the Rust code generated by bindgen to be a set of extern declarations in a C or C++ header. No more, no less.

Rust has native, easy interop with C libraries. It would be an odd restriction to force Rust programs to go through the serialization APIs if the yarp maintainers do not have that expectation for C code that consumes this library.

turboladen replied:
> Alright, then I think probably only parsing (yp_parse and its dependencies, including all the node structs) should be exposed to Rust, maybe lexing too if that's useful.

I think that, along with getting diagnostic info ("why did parsing fail?") would work for me. For my part, I don't think I have a use for lexing at this point.

turboladen replied:
> BTW do you really have some code/prototype using this? 

I have 3 repos of LSP prototypes, going back 2 years or so, but they're all a bit of a mess. :D Everything is pretty straightforward until I want to account for multiple definitions of something (class, method, etc) that exist in different files/workspaces/dependencies. But like I mentioned above, my bare-bones needs are:

- a `parse()` function (obviously :) ), which returns some collection of nodes (AST, CST, whatever) and diagnostic info (so I can report back to the client where code in a file is invalid)
- node info (type, location of the whole node) for each node in the collection
- the ability to navigate (walk, traverse, whatever) the collection would be nice. Ideally, it'd be really easy to look up nodes that are relevant to the node where the cursor is: if the cursor is inside an instance method on a class, I want to know all the things that are in scope. Surely there are lots of ways to solve this, but I find myself wanting to implement that functionality ("give me all the nodes in scope X") on the collection of nodes that I've parsed. Probably out of `yarp`'s scope, but thought I'd throw that out there. :)

I'm guessing it'd also be out of `yarp`'s scope, but I do like how in tree-sitter, you can "edit" a parse tree in place: https://docs.rs/tree-sitter/0.20.10/tree_sitter/struct.Tree.html#method.edit. I've assumed, of course, that this functionality helps with efficiency, but never ran any benchmarks.

lopopolo commented on rust/yarp-sys/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+name = "yarp-sys"
+version = "0.1.0"
+edition = "2021"
+license-file = "../../LICENSE.md"
```
is it possible to add a `license` field with an SPDX expression so automated tools can parse the license? `license-file` is meant to be an escape hatch when it is otherwise impossible to specify the license via SPDX.

turboladen replied:
This makes sense to me, but after looking at the `LICENSE.md` file, I don't know what to populate `license` with...

lopopolo commented on rust/yarp-sys/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+]
+rust-version = "1.59.0"
+
+[lib]
+crate_type = ["staticlib", "rlib"]
```
I think this section can be removed. `staticlib` asks `rustc` to produce an `.a` archive. Why would we ask Rust to do that when I imagine the C toolchain in this repository can do the same thing?

turboladen replied:
🤦 Good point!

turboladen commented on rust/yarp-sys/tests/node_tests.rs:
```diff
@@ -0,0 +1,35 @@
+#[test]
+fn node_test() {
+    let mut parser = MaybeUninit::<yp_parser_t>::uninit();
+    let code = CString::new("class Foo; end").unwrap();
```
FWIW, this is not a question I could answer. IIRC, I saw instances of interior NULs in the C code somewhere, but I figured I'd play it safe and use a `CString` here.

For those unfamiliar with Rust's [CString](https://doc.rust-lang.org/std/ffi/struct.CString.html):

> A type representing an owned, C-compatible, nul-terminated string with no nul bytes in the middle.

turboladen commented on rust/yarp-sys/tests/parser_tests.rs:
```diff
@@ -0,0 +1,212 @@
+use std::{
+    cell::OnceCell,
```
I've removed the 1.59.0 MSRV (which was only initially specified because of a `bindgen` call that generates code that's dependent on 1.59); I removed that code and the MSRV.

turboladen commented on rust/yarp-sys/tests/parser_tests.rs:
```diff
@@ -0,0 +1,212 @@
+
+        // (Don't need to free `buffer` because `String` takes ownership.)
+        let string =
+            String::from_raw_parts(buffer.value as *mut u8, buffer.length, buffer.capacity);
```
Switched to use `slice`

turboladen commented on rust/yarp-sys/tests/parser_tests.rs:
```diff
@@ -0,0 +1,212 @@
+        });
+    }
+
+    static mut THING: OnceCell<u8> = OnceCell::new();
```
> I am concerned by the number of issues with the unsafe code in this crate.

I'm a bit confused by this--it's a crate entirely of FFI bindings, which require `unsafe` to use. The code calling these FFI functions is only in tests, so it's not like downstream crates will be effected by these calls. How else would you test FFI bindings without using `unsafe`?

turboladen replied:
> `OnceCell` provides interior mutability and can be used as a regular `static`.

Thanks for the pointer here; I've only ever used `lazy_static`, so this API is new to me. I'll fix.

turboladen replied:
Hmm... if I remove the `mut`, then I can no longer call `get_mut()` (line 136). What am I missing here?

```
error[E0596]: cannot borrow immutable static item `THING` as mutable
   --> tests/parser_tests.rs:136:9
    |
136 |         THING.get_mut().map(|v| {
    |         ^^^^^^^^^^^^^^^ cannot borrow as mutable
```

lopopolo replied:
I'm not quite sure how you are intending to use `OnceCell`. It is a type intended for lazy initialization. If you want to use it to check your callback is called, I would use `THING.set(1).unwrap()` in the callback and `THING.get().unwrap()` in the assertion.

lopopolo replied:
My rule of thumb is that the `static mut` construct is meant to be used for like, implementing interrupt handlers in an embedded context. For basically anything else, using `static mut` means I'm holding it wrong.

turboladen replied:
> I'm not quite sure how you are intending to use `OnceCell`. It is a type intended for lazy initialization. If you want to use it to check your callback is called, I would use `THING.set(1).unwrap()` in the callback and `THING.get().unwrap()` in the assertion.
> 
> The real issue here though is you are using a non-thread safe type in a static.
> 
> You should be using https://doc.rust-lang.org/stable/std/sync/struct.OnceLock.html.

Ah yeah, I switched to `OnceLock` locally after testing out removing the `mut`. And yeah, really I only care about testing that the callback was called, so yeah it makes sense to switch to calling `set()` in the callback. Thanks for the tip.

turboladen commented on rust/yarp-sys/tests/parser_tests.rs:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    static mut THING: OnceCell<u8> = OnceCell::new();
+    unsafe { THING.set(42).unwrap() };
```
I removed, but I'm not sure how to mutate the inner variable without declaring `THING` as `mut`. `get_mut()` requires `&mut self`, not `&self`: https://doc.rust-lang.org/std/sync/struct.OnceLock.html#method.get_mut

turboladen commented:
> I'd like to see the CI on this crate expanded to:
> 
>     * Deny warnings.
> 
I get it, but I really hate working in crates that are set up like this; doing active development is such a pain.

>     * Run clippy.
> 
I'd already done this as part of my initial PR, but will add a GH job for this.

>     * Run the tests under leak sanitizer using `rustc`'s `-Z sanitzer=leak` option.
> 
>     * Run the tests under address sanitizer using `rustc`'s `-Z sanitizer=address` option.

This doesn't seem to me to add much value, and will add to the GH action time. I'd say, 100% do this in a crate that consumes these bindings, but just in the Rust tests here? 

I mean, sanitizers aside, what value do the tests provide? They're certainly not validating the C API (and IMO, they shouldn't--that should be done by C/Ruby tests); I'd say they're just validating we can call `yp_` functions as we'd expect, and thus don't actually really provide much value outside maybe serving as examples of how to string together some related `yp_` functions. And in fact, they're probably not even great at that since I don't think any of them are actually realistic, so to speak.

lopopolo commented:
> > I'd like to see the CI on this crate expanded to:
> > ```
> > * Deny warnings.
> > ```
> 
> I get it, but I really hate working in crates that are set up like this; doing active development is such a pain.

I don't like doing local development with deny warnings either. One is able to deny warnings in CI without adding `#![deny(warnings)]` to `lib.rs`.

This configuration can be applied only in CI by setting the `RUSTFLAGS` env variable to `-D warnings` in CI. GitHub Actions makes it easy to set environment configuration like this for an entire workflow.

turboladen commented on rust/yarp-sys/tests/parser_tests.rs:
```diff
@@ -0,0 +1,212 @@
+        width: usize,
+    ) -> *mut yp_encoding_t {
+        let c_name = CStr::from_ptr(name);
+        assert_eq!(c_name.to_string_lossy(), "meow");
+        assert_eq!(width, 4);
```
I suppose I've never encountered asserting inside an `extern "C" fn`... The docs there seem a bit confusing to me. On initial read, I'm inclined to change this, but then https://doc.rust-lang.org/beta/nomicon/ffi.html#panic-can-be-stopped-at-an-abi-boundary makes it sound like my assert calls would be fine. And, in fact, if I change one of those assertions, I get a regular ol' Rusty failure:

```
failures:

---- encoding_decode_test stdout ----
thread 'encoding_decode_test' panicked at 'assertion failed: `(left == right)`
  left: `"meow"`,
 right: `"pants"`', tests/parser_tests.rs:181:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

...which (admittedly, naïvely) seems like it's ok?

lopopolo replied:
this behavior is currently being changed as part of the rollout of the `C-unwind` ABI. In future Rust, this will turn into an `abort(3)` which will forcibly terminate the test process.

turboladen replied:
Solved by using another icky `OnceLock`.

lopopolo commented:
> I'd already done this as part of my initial PR, but will add a GH job for this.
> 
> > ```
> > * Run the tests under leak sanitizer using `rustc`'s `-Z sanitzer=leak` option.
> > 
> > * Run the tests under address sanitizer using `rustc`'s `-Z sanitizer=address` option.
> > ```
> 
> This doesn't seem to me to add much value, and will add to the GH action time. I'd say, 100% do this in a crate that consumes these bindings, but just in the Rust tests here?
> 
> I mean, sanitizers aside, what value do the tests provide?

There is a non trivial amount of unsafe Rust code in the tests for this crate and the sanitizers I mentioned above would have caught several misuses of `from_raw_parts` present in previous iterations of this PR.

I personally would prefer all the help I can get when writing unsafe Rust.

My understanding is that GitHub Actions is free for open source organizations like @ruby. If blocking CI time is a concern, including the sanitizer jobs as additional jobs in the `rust` workflow should allow the checks to be non-blocking by not marking them as required.

turboladen commented:
> One is able to deny warnings in CI without adding `#![deny(warnings)]` to `lib.rs`.

Ah, I misunderstood your comment. Good call. I'll change.

turboladen commented on README.md:
```diff
@@ -44,10 +46,10 @@ The repository contains the infrastructure for both a shared library (librubypar
 
-* A C99 compiler
-* autotools (autoconf, automake, libtool)
-* make
-* Ruby 3.3.0-preview1 or later
+- A C99 compiler
+- autotools (autoconf, automake, libtool)
+- make
+- Ruby 3.3.0-preview1 or later
```
I actually hadn't realized I'd made these changes; I have my IDE setup to run `prettier` on save. Happy to revert if anyone cares.

eregon replied:
Yes, no unneeded changes please, it's already big enough of a PR

turboladen replied:
Reverted

eregon commented on rust/yarp-sys/tests/unescape_tests.rs:
```diff
@@ -0,0 +1,79 @@
+
+use yarp_sys::{
+    yp_list_empty_p, yp_list_init, yp_list_t, yp_parser_init, yp_parser_t, yp_string_t,
+    yp_string_t_type, yp_unescape_calculate_difference, yp_unescape_manipulate_string,
+    yp_unescape_type_t,
+};
+
+#[test]
+fn unescape_manipulate_string_test() {
```
`yp_unescape_` is internal stuff, I don't think it makes sense to expose that to Rust.
(it's exposed under YARP::Debug for YARP internal testing but that's it, it's not meant to be used by anything but YARP itself)

turboladen replied:
My intent was to follow the `YP_EXPORTED_FUNCTION`s, but it sounds like that was a mistake. Is there a better way to deduce what's in the public API?

kddnewton commented:
This thread is getting difficult to parse through because of the number of comments. I think at this point I'm pretty satisfied with the work that has been done, and I'm happy to merge this as is and then iterate on the `main` branch going forward before we release to cargo. @turboladen if you don't mind just checking out the CI failures, once this is passing I'm happy to merge.

turboladen commented:
I think this latest push should fix all the tests. Sorry for the churn. There are still 3 failures total in the sanitizer jobs, pointing to memory leaks in the tests. Normally I'd want to fix those before merging, but given a) the length this PR has been open, and b) the fact that I'm on a Mac and can't easily run tests using the sanitizers (normally I set up a small docker container and run inside those; IMO it'd be nice to commit that setup for other devs on Mac that end up here), personally I'd like to merge this, then do a subsequent PR for addressing the leaks in the tests. If, of course, you'd like me to address those before merging, that's totally understandable, just let me know.

turboladen commented:
> I don't want to merge if the CI is red and it's leaking. I went through the leaks and I believe I've found them all. I've added a patch file here you can apply that should fix them all

Thanks @kddnewton! Patch applied, rebased, and pushed.

kddnewton commented:
@turboladen I think it may have lost some of your more recent changes, like where you caught it up to `0.7.0`.

kddnewton commented:
Ahh man, yeah that function doesn't exist anymore. It's a macro that's `YP_LIST_EMPTY` that just sets everything to 0. If you can calloc it should be fine.

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_type("yp_diagnostic_t")
+        .allowlist_type("yp_encoding_changed_callback_t")
+        .allowlist_type("yp_encoding_decode_callback_t")
+        .allowlist_type("yp_memsize_t")
```
I don't think this needs to be exposed.

eregon replied:
This comment meant to only quote one line, I think the concurrent git push messed it up.
Basically assume I'm only talking about the last line of every comment unless mentioned otherwise.

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_type("yp_comment_t")
+        .allowlist_type("yp_diagnostic_t")
+        .allowlist_type("yp_encoding_changed_callback_t")
+        .allowlist_type("yp_encoding_decode_callback_t")
```
This is only needed for encodings YARP does not understand. I would expect this is extremely rare, so probably unnecessary to expose.

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_type("yp_buffer_t")
+        .allowlist_type("yp_comment_t")
+        .allowlist_type("yp_diagnostic_t")
+        .allowlist_type("yp_encoding_changed_callback_t")
```
I don't think this needs to be exposed.
The encoding can be read after yp_parse(), I don't there is a need to know it before.

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .sort_semantically(true)
+        // Structs
+        .allowlist_type(r#"^yp_\w+_node_t"#)
+        .allowlist_type("yp_buffer_t")
```
I don't think this needs to be exposed, unless you want to use `yp_parse_serialize()`.

eregon replied:
But you are using and testing that, so all good.

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_type("yp_list_t")
+        .allowlist_type("yp_node_t")
+        .allowlist_type("yp_node_type")
+        .allowlist_type("yp_parser_t")
```
This struct has way too much internal impl details, but unfortunately there are not enough functions yet to access it like an opaque pointer or so, e.g. for comments/errors/warnings, encoding and maybe `start`/`end`.
So until then it has to be exposed to Rust.

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_type("yp_node_t")
+        .allowlist_type("yp_node_type")
+        .allowlist_type("yp_parser_t")
+        .allowlist_type("yp_pack_size")
```
It's already in `.rustified_non_exhaustive_enum` below, is it necessary here?

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_type("yp_node_type")
+        .allowlist_type("yp_parser_t")
+        .allowlist_type("yp_pack_size")
+        // TODO: Commenting this because I can't figure out how to get bindgen to generate the
+        // inner-unions nicely. Hand-rolling this in src/lib.rs for now.
+        // .allowlist_type("yp_string_t")
```
This changed, now `yp_string_t` is a simple struct, so probably can uncomment and remove the comment now

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_type("yp_token_type_t")
+        // Enums
+        .rustified_non_exhaustive_enum("yp_comment_type_t")
+        .rustified_non_exhaustive_enum("yp_context_t")
+        .rustified_non_exhaustive_enum("yp_heredoc_indent_t")
+        .rustified_non_exhaustive_enum("yp_heredoc_quote_t")
+        .rustified_non_exhaustive_enum("yp_lex_mode_t")
+        .rustified_non_exhaustive_enum("yp_lex_state_t")
```
I don't think any of these need to be exposed.

eregon replied:
This comment means all 5 lines quoted

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .rustified_non_exhaustive_enum("yp_pack_type")
+        .rustified_non_exhaustive_enum("yp_pack_variant")
+        .rustified_non_exhaustive_enum("yp_token_type")
+        .rustified_non_exhaustive_enum("yp_unescape_type_t")
```
I don't think this needs to be exposed.

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_function("yp_list_empty_p")
+        .allowlist_function("yp_list_free")
+        .allowlist_function("yp_list_init")
+        .allowlist_function("yp_node_memsize")
```
I don't think this needs to be exposed.

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_function("yp_parse_serialize")
+        .allowlist_function("yp_parser_free")
+        .allowlist_function("yp_parser_init")
+        .allowlist_function("yp_parser_register_encoding_changed_callback")
+        .allowlist_function("yp_parser_register_encoding_decode_callback")
```
I don't think this needs to be exposed (see above).

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_function("yp_parser_register_encoding_changed_callback")
+        .allowlist_function("yp_parser_register_encoding_decode_callback")
+        .allowlist_function("yp_prettyprint")
+        .allowlist_function("yp_regexp_named_capture_group_names")
```
I don't think this needs to be exposed, i.e. I think the name are already exposed in the resulting AST nodes

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_function("yp_prettyprint")
+        .allowlist_function("yp_regexp_named_capture_group_names")
+        .allowlist_function("yp_serialize")
+        .allowlist_function("yp_size_to_native")
```
I don't think this needs to be exposed

eregon commented on rust/yarp-sys/build.rs:
```diff
@@ -0,0 +1,146 @@
+        .allowlist_function("yp_string_list_init")
+        .allowlist_function("yp_string_list_free")
+        .allowlist_function("yp_token_type_to_str")
+        .allowlist_function("yp_unescape_calculate_difference")
+        .allowlist_function("yp_unescape_manipulate_string")
```
I don't think this needs to be exposed

eregon reviewed (commented):
I went through `rust/yarp-sys/build.rs` and trying to note what's internal or what's not necessary to expose for parsing and pack format parsing.

eregon commented on rust/yarp-sys/tests/unescape_tests.rs:
```diff
@@ -0,0 +1,79 @@
+};
+
+#[test]
+fn unescape_manipulate_string_test() {
```
I think this file should be removed because it uses internal APIs that I don't think are of any use if you are using `yp_parse()` (it will already unescape things for you)

kddnewton commented:
@turboladen sorry, another function signature changed out from under you. I'll get you a patch ready for that.

@eregon I agree this exposes a lot of stuff, but in the interest of getting this PR merged, I'm overlooking that in the hopes to change it on `main`. It's already very difficult to read through this whole thread.

eregon commented:
Yeah I'm OK if this is cleaned up after merge. It should be done somewhat soon, because I don't think we want to spend much effort updating Rust files if some of it goes away like the unescape test.

eregon commented:
I think the one thing I'd like done in this PR is removing `rust/yarp-sys/tests/unescape_tests.rs`, there is no point to try to fix it if the corresponding API is internal and there is no real value for it to be public or exposed to Rust.

eregon commented:
And the fact the CI specifically fails on that is good coincidence, relying on internals is useless churn, better remove that test file now than fix it for nothing.

kddnewton commented:
Actually yeah @turboladen if you want to just remove `unescape_tests.rs` it should be passing and then let's merge.

turboladen commented:
> Actually yeah @turboladen if you want to just remove `unescape_tests.rs` it should be passing and then let's merge.

Will do.

turboladen commented:
@eregon for my part, I've got some PR fatigue here and consequently would like to get this whole baseline merged in; I think your feedback on the API surface area is great and should be addressed sooner than later. Just my two cents on the matter though.

kddnewton commented:
@turboladen thank you so much for you patience. I think I can take it from here. I really appreciate all the effort you put in! If you aren't too burned out and feel like helping out on the main `yarp` crate (not just `yarp-sys`) I think we can start on that soon.

turboladen commented:
> @turboladen thank you so much for you patience. I think I can take it from here. I really appreciate all the effort you put in! If you aren't too burned out and feel like helping out on the main `yarp` crate (not just `yarp-sys`) I think we can start on that soon.

My pleasure! I do wish I'd have done more work up front to minimize the churn, but oh well. Goes to show how nice it is to have a community to help make things better. But yeah, after a breather, I'd love to stick around and help out with `yarp`-to-be and `yarp-sys`.
